File: Leetcode/Typescript/2145. Count the Hidden Sequences/solution.ts
// https://leetcode.com/problems/count-the-hidden-sequences/description

function numberOfArrays(
  differences: number[],
  lower: number,
  upper: number
): number {
  let prefixSum = 0;
  let minNum = 0;
  let maxNum = 0;

  // Compute prefix sums and track the min and max values
  for (const diff of differences) {
    prefixSum += diff;

    minNum = Math.min(minNum, prefixSum);
    maxNum = Math.max(maxNum, prefixSum);
  }

  const minStart = lower - minNum;
  const maxStart = upper - maxNum;

  // Calculate how many valid starting values arr[0] can take
  // so that all elements stay within [lower, upper]
  const count = maxStart - minStart + 1;

  // If the count is negative, it means no valid sequence exists
  return count < 0 ? 0 : count;
}

File: Leetcode/Python/2145. Count the Hidden Sequences/solution.py
# https://leetcode.com/problems/count-the-hidden-sequences/description

class Solution:
    def numberOfArrays(self, differences: List[int], lower: int, upper: int) -> int:
        prefix_sum = 0 
        range_min = range_max = 0  

        # Compute prefix sums and track the min and max values
        for diff in differences:
            prefix_sum += diff
            range_min = min(range_min, prefix_sum) 
            range_max = max(range_max, prefix_sum)  

        sequence_range = range_max - range_min

        # Calculate how many valid starting values arr[0] can take
        # so that all elements stay within [lower, upper]
        hidden_sequences = upper - lower - sequence_range + 1

        # If the count is negative, it means no valid sequence exists
        return 0 if hidden_sequences < 0 else hidden_sequences